packages/dds/matrix/src/test/time/sharedMatrix.spec.ts

```diff
@@ -0,0 +1,282 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import assert from "node:assert";
+
+import {
+	benchmark,
+	BenchmarkType,
+	isInPerformanceTestingMode,
+	type BenchmarkTimer,
+} from "@fluid-tools/benchmark";
+
+import { SharedMatrix } from "../../index.js";
+import { createLocalMatrix } from "../utils.js";
+
+describe("SharedMatrix execution time", () => {
+	// The value to be set in the cells of the matrix.
+	const matrixValue = "cellValue";
+	// The test matrix's size will be 10*10, 100*100, 1000*1000.
+	const matrixSizes = isInPerformanceTestingMode
+		? [10, 100, 1000]
+		: // When not measuring perf, use a single smaller data size so the tests run faster.
+			[10];
+
+	// The number of operations to perform on the matrix.
+	const operationCounts = isInPerformanceTestingMode
+		? [10, 100, 1000]
+		: // When not measuring perf, use a single smaller data size so the tests run faster.
+			[5];
+
+	let localMatrix: SharedMatrix;
+	for (const matrixSize of matrixSizes) {
+		describe(`Size of ${matrixSize}*${matrixSize} SharedMatrix`, () => {
+			// Filter counts to ensure remove operation do not exceed matrixSize
+			const validRemoveCounts = operationCounts.filter((count) => count <= matrixSize);
+
+			// Insert-related tests that are not limited by matrixSize
+			for (const count of operationCounts) {
+				// Test the execute time of the SharedMatrix for inserting a column in the middle for a given number of times.
+				benchmark({
+					type: BenchmarkType.Measurement,
+					title: `Insert a column in the middle ${count} times`,
+					benchmarkFnCustom: async <T>(state: BenchmarkTimer<T>) => {
+						let duration: number;
+						do {
+							// Since this setup one collects data from one iteration, assert that this is what is expected.
+							assert.equal(state.iterationsPerBatch, 1);
+							// Setup
+							localMatrix = createLocalMatrix({
+								id: "testLocalMatrix",
+								size: matrixSize,
+								initialValue: matrixValue,
+							});
```

Comment:
Is there not a way we can extract the setup step to ensure we don't include this in the measurements? Especially for large dense matrixes, this setup time will likely dominate the execution time, which will make it difficult to reason about the measurements we record.

Reply:
Focusing on `Set a 3-character string in N cells` for a sec, on size 1000, I'd be curious to see repeated runs of it and see how different the numbers are. With N=10, the screenshot shows a measurement of 6,000ns with ~100% margin of error, which feels pretty bad... but if repeated runs consistently measure around 6,000ns, even with high margins of error, the numbers might still be useful? @chentong7 can you do several runs of just that test and only on size 1000 to see how they compare?